File: tools/tirtc-device-token/java/TirtcSigningTest.java
import java.io.IOException;
import java.nio.file.*;
import java.time.OffsetDateTime;
import java.util.List;
import java.util.Map;

/**
 * Tests for TirtcSigning — validates against shared test-vectors.json.
 *
 * Compile and run:
 *   javac TirtcSigning.java TirtcSigningTest.java
 *   java TirtcSigningTest
 */
public class TirtcSigningTest {

    // Minimal JSON parser for test-vectors.json (avoids external dependencies)
    private static final OffsetDateTime SIGNING = OffsetDateTime.parse("2024-01-15T12:00:00Z");

    public static void main(String[] args) throws Exception {
        String json = Files.readString(Path.of("../test-vectors.json"));
        int passed = 0;

        // Deterministic
        Map<String, String> h1 = TirtcSigning.signRequest("access123", "secret", "app456", "POST", "/v1/token/wxvoip", "{\"k\":\"v\"}", "", SIGNING);
        Map<String, String> h2 = TirtcSigning.signRequest("access123", "secret", "app456", "POST", "/v1/token/wxvoip", "{\"k\":\"v\"}", "", SIGNING);
        check(h1.get("Authorization").equals(h2.get("Authorization")), "Deterministic");
        passed++;

        // Different secrets
        h1 = TirtcSigning.signRequest("access", "secret1", "app", "POST", "/path", "body", "", SIGNING);
        h2 = TirtcSigning.signRequest("access", "secret2", "app", "POST", "/path", "body", "", SIGNING);
        check(!h1.get("Authorization").equals(h2.get("Authorization")), "Different secrets");
        passed++;

        // Contains algorithm
        Map<String, String> h = TirtcSigning.signRequest("access", "secret", "app", "POST", "/v1/token/wxvoip", "{}", "", SIGNING);
        check(h.get("Authorization").startsWith(TirtcSigning.TGV1_ALG), "Contains algorithm");
        passed++;

        // Required headers
        h = TirtcSigning.signRequest("access", "secret", "app", "POST", "/v1/token/wxvoip", "{}", "", SIGNING);
        check(!h.get("X-Tg-Algorithm").isEmpty(), "Required X-Tg-Algorithm");
        check(!h.get("Content-Type").isEmpty(), "Required Content-Type");
        passed++;

        // GET no content-type
        h = TirtcSigning.signRequest("access", "secret", "app", "GET", "/v1/devices", "", "status=online", SIGNING);
        check(!h.containsKey("Content-Type"), "GET no Content-Type");
        check(!h.containsKey("Content-Length"), "GET no Content-Length");
        passed++;

        // URI trailing slash
        h1 = TirtcSigning.signRequest("access", "secret", "app", "POST", "/path/", "{}", "", SIGNING);
        h2 = TirtcSigning.signRequest("access", "secret", "app", "POST", "/path", "{}", "", SIGNING);
        check(h1.get("Authorization").equals(h2.get("Authorization")), "URI trailing slash");
        passed++;

        // --- Cross-language test vectors via manual verification ---
        // Vector 0: POST with body
        h = TirtcSigning.signRequest("test-access-key-123", "test-secret-456", "app-789",
                "POST", "/v2/device/info", "{\"device_id\":\"TESTDEVICE01\"}", "", SIGNING);
        check("2144f990e3b387b300f35c2222162ee10186cc35884e9b61b3194092447f3e2f"
                .equals(extractSignature(h.get("Authorization"))), "Vector 0 POST body");
        check("bf5afebe060d14c053ad5ca8ae574b300b305ea382da713600df86a93508f478"
                .equals(h.get("X-Tg-Content-Sha256")), "Vector 0 payload hash");
        check("28".equals(h.get("Content-Length")), "Vector 0 content-length");
        passed++;

        // Vector 1: POST empty body
        h = TirtcSigning.signRequest("test-access-key-123", "test-secret-456", "app-789",
                "POST", "/v2/user/login", "", "", SIGNING);
        check("e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855"
                .equals(h.get("X-Tg-Content-Sha256")), "Vector 1 empty body hash");
        check("0".equals(h.get("Content-Length")), "Vector 1 content-length");
        passed++;

        // Vector 2: GET with query
        h = TirtcSigning.signRequest("test-access-key-123", "test-secret-456", "app-789",
                "GET", "/v2/device/server/connection", "", "device_id=TESTDEVICE01&platform=web", SIGNING);
        check("b8e12cc829768a466a1e6820099a7b1172357327e6c2a7c692c3f78a1c1a057f"
                .equals(extractSignature(h.get("Authorization"))), "Vector 2 GET with query");
        check(!h.containsKey("Content-Type"), "Vector 2 no content-type");
        passed++;

        // Vector 3: GET plus in query
        h = TirtcSigning.signRequest("test-access-key-123", "test-secret-456", "app-789",
                "GET", "/v2/search", "", "q=hello+world", SIGNING);
        check("faf82b0dec7a2430b99b568192a53d042f66febf49a0f29d2244132c4dae4258"
                .equals(extractSignature(h.get("Authorization"))), "Vector 3 plus query");
        passed++;

        // Vector 4: PUT
        h = TirtcSigning.signRequest("test-access-key-123", "test-secret-456", "app-789",
                "PUT", "/v2/device/attrs", "{\"attrs\":{\"wakeup\":\"on\"}}", "", SIGNING);
        check("56090deedb1ffc6db164a21d4e75a383e1c3f1a55cd273a852167c0e282e96a1"
                .equals(extractSignature(h.get("Authorization"))), "Vector 4 PUT");
        check("25".equals(h.get("Content-Length")), "Vector 4 content-length");
        passed++;

        // Vector 5: DELETE
        h = TirtcSigning.signRequest("test-access-key-123", "test-secret-456", "app-789",
                "DELETE", "/v2/user/12345", "", "", SIGNING);
        check("62e5e2d93d5f61b880c7bea4eb5a3931f0ed465b069382ff31fe4eef82677d5a"
                .equals(extractSignature(h.get("Authorization"))), "Vector 5 DELETE");
        passed++;

        // Vector 6: URI trailing slash
        h = TirtcSigning.signRequest("test-access-key-123", "test-secret-456", "app-789",
                "POST", "/v2/device/info/", "{\"device_id\":\"TESTDEVICE01\"}", "", SIGNING);
        check("2144f990e3b387b300f35c2222162ee10186cc35884e9b61b3194092447f3e2f"
                .equals(extractSignature(h.get("Authorization"))), "Vector 6 trailing slash");
        passed++;

        // Vector 7: root path
        h = TirtcSigning.signRequest("test-access-key-123", "test-secret-456", "app-789",
                "GET", "/", "", "action=ping", SIGNING);
        check("67a9a490b68813c8df7a743325ae5336fbbd979a3fafa929620a34da7861f218"
                .equals(extractSignature(h.get("Authorization"))), "Vector 7 root path");
        passed++;

        System.out.println("PASS: all tests passed (" + passed + " tests, 8 vectors verified)");
    }

    private static String extractSignature(String auth) {
        int idx = auth.lastIndexOf("Signature=");
        return idx >= 0 ? auth.substring(idx + "Signature=".length()) : "";
    }

    private static void check(boolean condition, String msg) {
        if (!condition) {
            throw new AssertionError("FAIL: " + msg);
        }
    }
}

File: tools/tirtc-device-token/javascript/tirtc_signing.test.js
"use strict";

const fs = require('fs');
const path = require('path');
const { signRequest, TGV1_ALG } = require('./tirtc_signing');

function loadTestVectors() {
  const p = path.join(__dirname, '..', 'test-vectors.json');
  return JSON.parse(fs.readFileSync(p, 'utf-8')).vectors;
}

const vectors = loadTestVectors();
const signing = new Date('2024-01-15T12:00:00Z');

// Run basic tests
function assertEqual(actual, expected, msg) {
  if (actual !== expected) {
    throw new Error(`${msg}\n  expected: ${expected}\n  actual:   ${actual}`);
  }
}

// Deterministic
{
  const h1 = signRequest('access123', 'secret', 'app456', 'POST', '/v1/token/wxvoip', '{"k":"v"}', '', signing);
  const h2 = signRequest('access123', 'secret', 'app456', 'POST', '/v1/token/wxvoip', '{"k":"v"}', '', signing);
  assertEqual(h1.Authorization, h2.Authorization, 'Deterministic');
}

// Different secrets -> different sig
{
  const h1 = signRequest('access', 'secret1', 'app', 'POST', '/path', 'body', '', signing);
  const h2 = signRequest('access', 'secret2', 'app', 'POST', '/path', 'body', '', signing);
  if (h1.Authorization === h2.Authorization) {
    throw new Error('Different secrets must produce different sigs');
  }
}

// Contains algorithm prefix
{
  const h = signRequest('access', 'secret', 'app', 'POST', '/v1/token/wxvoip', '{}', '', signing);
  assertEqual(h.Authorization.startsWith(TGV1_ALG), true, 'Contains algorithm');
}

// Required headers for POST
{
  const h = signRequest('access', 'secret', 'app', 'POST', '/v1/token/wxvoip', '{}', '', signing);
  ['X-Tg-Algorithm', 'X-Tg-Date', 'X-Tg-App-Id', 'X-Tg-Content-Sha256', 'Content-Type'].forEach(name => {
    if (!h[name]) throw new Error(`Header ${name} should not be empty`);
  });
}

// GET without content-type
{
  const h = signRequest('access', 'secret', 'app', 'GET', '/v1/devices', '', 'status=online', signing);
  if (h['Content-Type'] || h['Content-Length']) {
    throw new Error('GET should not have Content-Type/Length');
  }
}

// URI trailing slash
{
  const h1 = signRequest('access', 'secret', 'app', 'POST', '/path/', '{}', '', signing);
  const h2 = signRequest('access', 'secret', 'app', 'POST', '/path', '{}', '', signing);
  assertEqual(h1.Authorization, h2.Authorization, 'URI trailing slash');
}

// --- Cross-language test vectors ---
vectors.forEach((v, i) => {
  const h = signRequest(v.accessKey, v.accessSecret, v.appId, v.method, v.uriPath, v.body, v.rawQuery, signing);
  Object.keys(v.expected).forEach(key => {
    assertEqual(
      h[key] || '',
      v.expected[key],
      `[${v.description}] header '${key}' mismatch`
    );
  });
});

console.log(`PASS: all tests passed (${vectors.length} vectors verified)`);
